examples/ode-integrator.dx

```diff
@@ -1,60 +1,180 @@
+alpha = [1. / 5., 3. / 10., 4. / 5., 8. / 9., 1., 1.]
+
+-- Todo: Write as triangular array when concrete syntax supports it.
+beta' = [[1. / 5., 0., 0., 0., 0., 0.],
+       [3. / 40., 9. / 40., 0., 0., 0., 0.],
+       [44. / 45., -56. / 15., 32.0 / 9., 0., 0., 0.],
+       [19372. / 6561., -25360. / 2187., 64448. / 6561., -212. / 729., 0., 0.],
+       [9017./3168., -355./33., 46732./5247., 49./176., -5103./18656., 0.],
+       [35. / 384., 0., 500. / 1113., 125./192., -2187./6784., 11./84.]]
+-- Convert to triangular array.
+beta = for i. for j:(..i). beta'.i.((ordinal j)@(Fin 6))
+
+c_sol = [35. /384., 0., 500. /1113., 125. /192., -2187. /6784., 11. / 84., 0.]
+c_error = [35. / 384. - 1951. / 21600., 0., 500. / 1113. - 22642. / 50085.,
+           125. / 192. - 451. / 720., -2187. / 6784. + 12231. / 42400.,
+           11. / 84. - 649. / 6300., -1. / 60.]
+
+def runge_kutta_step (_:VSpace v) ?=> (func:v->Time->v)
+    (z0:v) (f0:v) (t0:Time) (dt:Time) : (v & v & v & (Fin 7)=>v) =
+
+  runge_kutta_stage = \i func_evals.
+    ti = t0 + dt .* alpha.i
+    zi = z0 + dt .* dot beta.i func_evals
+    func zi ti
+
+  evals_init = for i. select (i == (0@_)) f0 zero
+
+  evals_filled = snd $ withState evals_init \func_evals. for i:(Fin 6).
+    cur_evals = for j:(..i). (get func_evals).((ordinal j)@_)
+    ft = runge_kutta_stage i cur_evals
+    func_evals!(((ordinal i) + 1)@_) := ft
+
+  z_last =  z0 + dt .* dot c_sol   evals_filled
+  z_last_error = dt .* dot c_error evals_filled
+  f_last = evals_filled.(6@_)
+  (z_last, f_last, z_last_error, evals_filled)
+
+def error_ratio (error_estimates:d=>Real) (rtol:Real) (atol:Real)
+                (z0:d=>Real) (z1:d=>Real) : Real =
+  err_tols = for i. atol + rtol * (max (abs z0.i) (abs z1.i))
+  err_ratios = error_estimates ./ err_tols
+  mean for i. sq err_ratios.i
+
+def optimal_step_size (last_step:Time) (mean_error_ratio:Real) : Time =
+  safety = 0.9
+  ifactor = 10.
+  dfactor = 0.2
+  order = 5.
+  dfactor = select (mean_error_ratio < 1.) 1. dfactor
+  err_ratio = sqrt mean_error_ratio
+  minfac = min ( (pow err_ratio (1. / order)) / safety) (1. / dfactor)
+  factor = max (1. / ifactor) minfac
+  select (mean_error_ratio == 0.) (last_step * ifactor) (last_step / factor)
+
+
+def odeint (func: d=>Real -> Time -> d=>Real)
+           (z0: d=>Real) (t0: Time) (times: n=>Time) : n=>d=>Real =
+  -- Adaptive stepsize (Dormand-Prince) Runge-Kutta odeint implementation.
+  --  Args:
+  --    func: time derivative of the solution z at time t.
+  --    z0: the initial value for the state.
+  --    t: times for evaluation. values must be strictly increasing.
+  --  Returns:
+  --    Values of the solution at each time point in times.
+  rtol = 1.4e-8 -- relative local error tolerance for solver.
+  atol = 1.4e-8 -- absolute local error tolerance for solver.
+  max_iters = 10000
+
+  integrate_to_next_time = \iter init_carry.
+    target_t = times.iter
+
+    stopping_condition = \(_, _, t, dt, _, _).
```

Comment:
How about using extensible records for the solver state? That would make unpacking like this much more readable I think

Reply:
I tried making this change, and it does make this part more readable, but it made the rest of the function much more verbose.  However I'm not sure I have the hang of extensible records yet.  Perhaps we could merge it and then @hexahedria could update this example as a demo of extensible records?